### PR TITLE
175 replace input output@main

### PR DESCRIPTION
**Pull request by BLAZEWIM**
# Pull Request

Fixes #[175](https://github.com/insightsengineering/teal.slice/issues/175)

also varname

### CODE REVIEW

**Comment by chlebowa** [comment]
Was #175 to add support for a character string in `dataname` or to change to only handle a character string?

**Comment by BLAZEWIM** [comment]
> Was #175 to add support for a character string in `dataname` or to change to only handle a character string?

Second one, change it.

**Comment by nikolas-burkoff** [comment]
Please merge into filterpanel_refactor@main once this has been merged 👍

File: R/FilterStates.R
Comment (nikolas-burkoff) [inline R/FilterStates.R:143]: if dataname is only a character then we don't need this? And on line 148

File: R/FilterStates.R
Comment (nikolas-burkoff) [inline R/FilterStates.R:104]: @gogonzo maybe another issue but on line 109 do you think we could do the following:

get_call = function(call_dataname = character(0)) {
  if (length(call_dataname) == 0) {
    call_dataname <- private$dataname
  } 
  checkmate::assert_character(call_dataname)  
  // replace private$dataname with call_dataname in the rest of this function
},

**Review by nikolas-burkoff — COMMENTED** [review]
A couple of small comments

File: R/FilterStates.R
Comment (nikolas-burkoff) [reply R/FilterStates.R:104]: I see this is likley to be too hard

File: tests/testthat/test-init_filter_state.R
Comment (chlebowa) [inline tests/testthat/test-init_filter_state.R:35]: remove

File: tests/testthat/test-init_filter_state.R
Comment (chlebowa) [inline tests/testthat/test-init_filter_state.R:45]: `dataname = "test"`

File: tests/testthat/test-init_filter_state.R
Comment (chlebowa) [inline tests/testthat/test-init_filter_state.R:49]: `dataname = "test"`

**Review by chlebowa — CHANGES_REQUESTED** [review]
Some assertions change with #171, wait for that to be merged (~2 hrs) before you apply.

File: R/FilterStates.R
Comment (BLAZEWIM) [reply R/FilterStates.R:143]: Reason is MAE - in this case it still can be a call. BUT - maybe it would be better to reverse it and create a custom get_call for FilterStatesMAE.R.

**Comment by BLAZEWIM** [comment]
> Some assertions change with #171, wait for that to be merged (~2 hrs) before you apply.

Ok, will wait, let me know.

File: R/FilterStates.R
Comment (nikolas-burkoff) [reply R/FilterStates.R:143]: at the very least add a comment here so when I look next time I see why it's like this :)

File: R/FilterState.R
Comment (donyunardi) [reply R/FilterState.R:54]: I assumed you were referring to making `dataname` argument as character class.

If we do this, we need to make sure that `$get_dataname()` is behaving properly.

https://github.com/insightsengineering/teal.slice/blob/aab1255ec1cbb6442f7d3cef1271eed873aec0a7/R/FilterState.R#L180-L186

If we treat `dataname` as character, then it will have backslash `\` when calling `$get_dataname`
```r
# default behavior
r$> filter_state$get_dataname()
[1] "data"

r$> deparse1(as.name("data"))
[1] "data"

# if we treat dataname as character
r$> deparse1("data")
[1] "\"data\""


```

Comment (chlebowa) [reply R/FilterState.R:54]: Yes, the `deparse` in the method should be removed.

Note the slash is just an esacape character because you get a string containing quote marks.

File: R/FilterStates.R
Comment (gogonzo) [reply R/FilterStates.R:147]: Why is custom call needed? If we have `dataname = character(1)` in all filter-states then we can still use abstract `get_call`, right? If you need to turn `character` to call please just use `str2lang()`

`language` is used only in `get_call` methods and we needed to convert dataname into character using `deparse1(private$input_dataname)` everywhere else. Point of the issue is to simplify the classes - let's keep this field as a `character` only and use `character` only to initialize the class.

Just a brief look how faster is turning from string to language. Comparison is still on the level of microseconds, but still shows that keeping `private$dataname = character(1)` and converting it to the `language` when using `get_call` is not a big deal.

```r
call <- quote(MAE[[xxx]])
microbenchmark::microbenchmark(
  str2lang("MAE[[xxx]]"),
  deparse1(call)
)
#                   expr    min     lq     mean median     uq    max neval
# str2lang("MAE[[xxx]]")  1.401  1.601  1.83901  1.701  1.801 12.201   100
#         deparse1(call) 15.500 15.901 17.13894 16.201 16.751 74.501   100
```

File: R/FilterStates-utils.R
Comment (chlebowa) [inline R/FilterStates-utils.R:9]:
```suggestion
#' @param dataname (`character(1)`)\cr
```

File: tests/testthat/test-FilterState.R
Comment (chlebowa) [inline tests/testthat/test-FilterState.R:42]: Amend test descriptions.

**Review by chlebowa — APPROVED** [review]
Almost there :+1:

File: R/FilterStateDate.R
Comment (BLAZEWIM) [reply R/FilterStateDate.R:69]: It can, the condition in it was a leftover, I will check whether we can get rid of it

File: R/FilterState.R
Comment (BLAZEWIM) [reply R/FilterState.R:174]: Thinking about it again, I will change it to empty string, this "NULL" looks strange. This was the default behavior previously (see modified tests_init_filter_states.R, `init_filter_state provides default values for varlabel, dataname` ), but not neccessary one I suppose.

File: R/FilterState.R
Comment (BLAZEWIM) [reply R/FilterState.R:464]: yes, but get_varname_prefixed() returns a call, so this is the place it stops being one

Comment (chlebowa) [reply R/FilterState.R:464]: Oh, I see!

File: R/FilteredDataset.R
Comment (gogonzo) [inline R/FilteredDataset.R:475]:
```suggestion
          logger::log_trace("FilteredDataset$initialized, dataname: { dataname }")
```

File: R/FilterStateDate.R
Comment (BLAZEWIM) [reply R/FilterStateDate.R]: I decided to keep the function, but change the use among the class to be consistent with the use of get_dataname and private$dataname

File: NEWS.md
Comment (gogonzo) [inline NEWS.md:26]: both classes and their constructor wrappers (init_filter_state, init_filter_states) are internal. They shouldn't be mentioned in the changelog

Comment (gogonzo) [reply NEWS.md:26]: There are also other NEWS about theese clases - we will remove them later

**Review by gogonzo — APPROVED** [review]
👍 Tested with the apps everything works. No occurrences of the input_ and output_dataname in the code. Calls are generated properly, logs looks good.

**Pull request merged**
